### PR TITLE
Supports advanced indexing of `MultiNestedTensor` (part 1)

### PR DESCRIPTION
**Pull request by weihua916**
- Supports row slicing
- Supports a column integer indexing.
- Refactor and code clean up.

Subsequent PRs:
- Supports column advanced indexing (slicing, List indexing)
- Supports concatenation along row and column.
- Increase test coverage.

### CODE REVIEW

File: test/data/test_multi_nested_tensor.py [hunk 1]
```diff
@@ -1,10 +1,22 @@
...
 import torch
+from torch import Tensor
 
 from torch_frame.data import MultiNestedTensor
 
 
+def assert_equal(tensor_mat: List[List[Tensor]],
+                 multi_nested_tensor: MultiNestedTensor):
+    assert len(tensor_mat) == multi_nested_tensor.num_rows
+    assert len(tensor_mat[0]) == multi_nested_tensor.num_cols
```
Comment (rusty1s) [inline test/data/test_multi_nested_tensor.py:13]: Can we also expose `shape`? I think `num_rows`, `num_cols` terminology for a tensor is quite confusing.

File: torch_frame/data/multi_nested_tensor.py [hunk 2]
```diff
@@ -78,85 +78,161 @@ def from_tensor_mat(
...
         index: Any,
     ) -> Union['MultiNestedTensor', Tensor]:
+        def col_index(
+                multi_nested_tensor: 'MultiNestedTensor',
+                index: Union[int, Tensor, List, slice]) -> 'MultiNestedTensor':
+            r"""Supports all types of column-level advanced indexing of input
+            :class:`MultiNestedTensor` object."""
+            if isinstance(index, int):
+                return multi_nested_tensor.single_index_select(index, dim=1)
+            elif isinstance(index, Tensor):
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:96]: There exists a bunch of non-tested code paths here.

File: torch_frame/data/multi_nested_tensor.py [hunk 3]
```diff
@@ -78,85 +78,161 @@ def from_tensor_mat(
     def __repr__(self) -> str:
         name = ' '.join([
             f"{self.__class__.__name__}(num_rows={self.num_rows},",
-            f"num_cols={self.num_cols},", f"device={self.values.device})"
+            f"num_cols={self.num_cols},", f"device={self.device})"
         ])
-
         return name
 
     def __getitem__(
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:85]: This method here is super complex. Can we refactor it?

File: torch_frame/data/multi_nested_tensor.py [hunk 4]
```diff
@@ -78,85 +78,161 @@ def from_tensor_mat(
...
+                return multi_nested_tensor.index_select(
+                    torch.tensor(index, device=multi_nested_tensor.device),
+                    dim=1)
+            elif isinstance(index, slice):
+                # TODO: Support this
+                return multi_nested_tensor.slice(index, dim=1)
+            else:
+                raise NotImplementedError
+
         if isinstance(index, Tuple):
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:110]: wouldn't it be easier if we just iterate over `index`, and then do:
```python
out = self
for i, value in enumerate(index):
    out = out.index_select(value, dim=i)

File: torch_frame/data/multi_nested_tensor.py [hunk 5]
```diff
@@ -78,85 +78,161 @@ def from_tensor_mat(
...
-    def _row_index_select(self, index: Tensor) -> Tensor:
-        index = index.clone()
-        # Support negative indices
-        neg_idx = index < 0
-        index[neg_idx] = self.num_rows + index[neg_idx]
-        if index.min() < 0 or index.max() >= self.num_rows:
-            raise IndexError("Index out of bounds!")
+            raise RuntimeError(f"Unsupported dim={dim} for index_select.")
+
+    def row_slice(self, slice: slice) -> 'MultiNestedTensor':
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:179]: this is known as narrow. Let's try to follow PyTorch terminology.

**Review by rusty1s — CHANGES_REQUESTED** [review]

File: torch_frame/data/multi_nested_tensor.py (hunk 5 above)
Comment (weihua916) [reply torch_frame/data/multi_nested_tensor.py:179]: I checked `narrow` (https://pytorch.org/docs/stable/generated/torch.narrow.html) but the arguments are different from `slice`.
It needs to pass `start` and `length`. There is no `step` and `length` is different from `end`.

File: torch_frame/data/multi_nested_tensor.py (hunk 3 above)
Comment (weihua916) [reply torch_frame/data/multi_nested_tensor.py:85]: Yes, will try.

File: torch_frame/data/multi_nested_tensor.py (hunk 2 above)
Comment (weihua916) [reply torch_frame/data/multi_nested_tensor.py]: Added a temporary test on `NotImplementedError`. Will resolve it soon.

File: torch_frame/data/multi_nested_tensor.py (hunk 5 above)
Comment (rusty1s) [reply torch_frame/data/multi_nested_tensor.py:112]: That's true, but you can compute end from start and length. Overall, I just think it is nicer when we inherit PyTorch functions here so that it feels like a normal tensor.

Comment (weihua916) [reply torch_frame/data/multi_nested_tensor.py:112]: I see. But it cannot model the step. Shall we add narrow by calling row_slice internally?

File: torch_frame/data/multi_nested_tensor.py [hunk 6]
```diff
@@ -120,83 +120,93 @@ def __repr__(self) -> str:
...
-            return MultiNestedTensor(num_rows=1, num_cols=self.num_cols,
-                                     values=values, offset=offset)
-        elif isinstance(index, Tensor) and index.ndim == 1:
-            return self.index_select(index, dim=0)
-        elif isinstance(index, List):
-            return self.index_select(
-                torch.tensor(index, device=self.values.device), dim=0)
+            if isinstance(index[0], int) and isinstance(index[1], int):
+                # Returns Tensor
+                return MultiNestedTensor.get_tensor(self, index[0], index[1])
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:136]: Any reason to make this static? Why not do `self.get_value(row: int, col: int)`?

File: torch_frame/data/multi_nested_tensor.py [hunk 7]
```diff
@@ -120,83 +120,93 @@ def __repr__(self) -> str:
...
+            else:
+                # Returns MultiNestedTensor
+                out = self
+                for dim, idx in enumerate(index):
+                    out = MultiNestedTensor.index(out, idx, dim)
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:141]: Here again. Would prefer:
```suggestion
                    out = out.select(out, idx, dim)
```

File: torch_frame/data/multi_nested_tensor.py [hunk 8]
```diff
@@ -120,83 +120,93 @@ def __repr__(self) -> str:
...
-        if dim == 0:
-            return self._row_index_select(index)
+            # Returns MultiNestedTensor
+            return MultiNestedTensor.index(self, index, dim=0)
+
+    def slice(self, slice: slice, dim: int) -> 'MultiNestedTensor':
+        if dim == 0 or dim == -3:
+            return self.row_slice(slice)
+        elif dim == 1 or dim == -2:
+            return self.col_slice(slice)
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:151]: Can we add an assert here in case `dim` is not well defined?

File: torch_frame/data/multi_nested_tensor.py (hunk 5 above)
Comment (rusty1s) [reply torch_frame/data/multi_nested_tensor.py:153]: Sounds good.

File: torch_frame/data/multi_nested_tensor.py [hunk 9]
```diff
@@ -120,83 +120,93 @@ def __repr__(self) -> str:
...
+
+    def row_slice(self, slice: slice) -> 'MultiNestedTensor':
+        r"""Slice along row (dim=0)"""
+        if slice.step is not None and slice.step > 1:
+            # If step is larger than 1, we reuse index_select along rows.
+            idx = torch.arange(self.num_rows, device=self.device)[slice]
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:157]: nit: `torch.tensor(range(self.num_rows)[slice])` would be a bit more memory friendly.

File: torch_frame/data/multi_nested_tensor.py [hunk 10]
```diff
@@ -210,6 +220,95 @@ def _row_index_select(self, index: Tensor) -> Tensor:
...
+                    if index < 0 or index > self.num_rows:
+                        raise IndexError("Row index out of bounds!")
+                else:
+                    if index < 0 or index >= self.num_rows:
+                        raise IndexError("Row index out of bounds!")
+            else:
+                if index < 0:
+                    index = index + self.num_cols
+                if is_slice_end:
+                    if index < 0 or index > self.num_cols:
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:285]:
```suggestion
                if is_slice_end and index < 0 or index > self.num_cols:
```

File: torch_frame/data/multi_nested_tensor.py [hunk 11]
```diff
@@ -210,6 +220,95 @@ def _row_index_select(self, index: Tensor) -> Tensor:
...
+                        raise IndexError("Column index out of bounds!")
+                else:
+                    if index < 0 or index >= self.num_cols:
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:288]:
```suggestion
                elif index < 0 or index >= self.num_cols:
```

File: torch_frame/data/multi_nested_tensor.py [hunk 12]
```diff
@@ -210,6 +220,95 @@ def _row_index_select(self, index: Tensor) -> Tensor:
...
+            torch.cumsum(diff, dim=0, out=offset[1:])
+            return MultiNestedTensor(num_rows=self.num_rows, num_cols=1,
+                                     values=values, offset=offset)
+        else:
+            raise RuntimeError(f"Unsupported dim={dim} for index_select.")
+
+    def _to_positive_index(
+        self,
+        index: Union[int, Tensor],
+        dim: int,
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:257]: There is a lot of duplicated logic here depending on `dim`. Might be easier to just add a `max_entries` argument here instead of `dim`?

File: torch_frame/data/multi_nested_tensor.py [hunk 13]
```diff
@@ -210,6 +220,95 @@ def _row_index_select(self, index: Tensor) -> Tensor:
...
+                    index = index + self.num_cols
+                if is_slice_end:
+                    if index < 0 or index > self.num_cols:
+                        raise IndexError("Column index out of bounds!")
+                else:
+                    if index < 0 or index >= self.num_cols:
+                        raise IndexError("Column index out of bounds!")
+        elif isinstance(index, Tensor):
+            assert not is_slice_end
+            index = index.clone()
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:292]: nit: In most cases, the clone here is unnecessary:
```python
neg_index = index[neg_idx]
if neg_index.numel() > 0:
    index = index.clone()
    index[neg_idx] = max_entries + neg_index

File: torch_frame/data/multi_nested_tensor.py [hunk 14]
```diff
@@ -210,6 +220,95 @@ def _row_index_select(self, index: Tensor) -> Tensor:
...
+                    raise IndexError("Row index out of bounds!")
+            else:
+                index[neg_idx] = self.num_cols + index[neg_idx]
+                if index.min() < 0 or index.max() >= self.num_cols:
+                    raise IndexError("Column index out of bounds!")
+        return index
+
+    @property
+    def device(self) -> torch.device:
+        return self.values.device
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:307]: Seems to already exist.

File: torch_frame/data/multi_nested_tensor.py [hunk 15]
```diff
@@ -230,17 +329,25 @@ def _apply(self, fn: Callable[[Tensor], Tensor]) -> 'MultiNestedTensor':
...
+
+    @property
+    def ndim(self) -> int:
         return 3
 
+    def dim(self) -> int:
+        return self.ndim
+
     @property
     def device(self) -> torch.device:
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:342]: Shall we also expose `dtype`?

File: torch_frame/data/multi_nested_tensor.py [hunk 16]
```diff
@@ -120,83 +120,93 @@ def __repr__(self) -> str:
...
-        index = index.clone()
-        # Support negative indices
-        neg_idx = index < 0
-        index[neg_idx] = self.num_rows + index[neg_idx]
-        if index.min() < 0 or index.max() >= self.num_rows:
-            raise IndexError("Index out of bounds!")
+            start_idx: int = self._to_positive_index(slice.start or 0, dim=0)
+            end_idx: int = self._to_positive_index(slice.stop or self.num_rows,
+                                                   dim=0, is_slice_end=True)
+            if start_idx == 0 and end_idx == self.num_rows:
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:163]:
```suggestion
            if start_idx <= 0 and end_idx >= self.num_rows:
```

**Review by rusty1s — APPROVED** [review]

File: torch_frame/data/multi_nested_tensor.py (hunk 13 above)
Comment (weihua916) [reply torch_frame/data/multi_nested_tensor.py:292]: That's a good point. Fixed.

File: torch_frame/data/multi_nested_tensor.py (hunk 12 above)
Comment (weihua916) [reply torch_frame/data/multi_nested_tensor.py:257]: Good point. Done.

File: torch_frame/data/multi_nested_tensor.py (hunk 8 above)
Comment (weihua916) [reply torch_frame/data/multi_nested_tensor.py:151]: We already raise RuntimeError.
